### PR TITLE
Packages

### PR DESCRIPTION
**Pull request by sp1ritCS**
I've created buildfiles for rpm (focus SuSE & Fedora), Arch & Flatpak. They could be upgraded later, if the need arises (.desktop or mime handling).  The rpm is currently in my [testing repository](https://build.opensuse.org/package/show/home:sp1rit:testing/clapper) on SuSE's buildservice. You can currently download signed rpms here: [software.opensuse.org//download.html?project=home:sp1rit:testing&package=clapper](https://software.opensuse.org//download.html?project=home%3Asp1rit%3Atesting&package=clapper) (I'd advice against adding the repository to your packagemanager, as this is just the testing repository).

I'm unsure if the flatpak is even a good idea to start with; due to it's permission system & no way to add additional codecs without rebuilding the package. If you (@Rafostar ) want a flatpak, you might need have flatpaks file-sandbox in mind. Currently it only plays videos in that are xdg-videos. 
I think the best idea if you want to use flatpak is, once you have a "menu screen" to scan xdg-videos for stuff to show and for everything else to use GNOME's filepicker portal, as this works with flatpak. (so that the package just has xdg-videos access, to try and minimize permissions)

### CODE REVIEW

**Comment by Rafostar** [comment]
@sp1ritCS 
Thank you very much for your work.

From GStreamer required plugins we need `glsinkbin` which usually comes as part of `gstreamer1-plugins-base` (are base plugins always available?) and `gtkglsink` which availability (package name) differs per distro. On Fedora it comes inside `gstreamer1-plugins-good-gtk` package, so you should probably add it to required plugins.

> I think the best idea if you want to use flatpak is, once you have a "menu screen" to scan xdg-videos for stuff to show and for everything else to use GNOME's filepicker portal, as this works with flatpak. (so that the package just has xdg-videos access, to try and minimize permissions)

I will check it out after I start creating file open menu. Do you know if `gstreamer-vaapi` is available on Flatpak? Publishing media player without HW accel seems kinda pointless nowadays.

**Comment by sp1ritCS** [comment]
I didn't have to install either of them, so at least on opensuse tumbleweed (gnome) both are install with gstreamer. I still added them, as I dont know of fedora (cant test currently) and rpm can figure out optimized dependencies thanks to it.
It was a bit confusing to, as `gstreamer1-plugins-base` conflicts with `gstreamer1-plugins-bad-free < 1.13` on fedora.

---

flatpak dosn't really have the concept of "dependency's", you have a runtime that has stuff, everything else you need to compile in during build. I've just checked (using gst-inspect-1.0 vaapi) but the org.gnome.Platform runtime has no gstreamer-vappi. Could I just add that as a module tho?

**Comment by Rafostar** [comment]
> Edit: just checked org.gnome.Totem also doen't use hw accel vaapi

After some checking I realised, that Totem does not support VAAPI at all. It was never working correctly with Clutter rendering, so the devs disabled it forcefully. In our case we use `gtkglsink` to render GTK video widget through OpenGL, so VAAPI works without problems.

For Flatpak modules we should probably add `gst-libav` and `gst-vaapi`. Maybe we could do the custom building of ffmpeg libav like they do it in flathub Totem: [https://github.com/flathub/org.gnome.Totem/commit/9b338fa84f6f51723360c7d5f2207c710a579036](https://github.com/flathub/org.gnome.Totem/commit/9b338fa84f6f51723360c7d5f2207c710a579036) With this approach we would have ALL the possible codecs available in flatpak package.

File: pkgs/arch/.SRCINFO
Comment (Rafostar) [inline pkgs/arch/.SRCINFO:9]: We also use `python3` in meson build.

File: pkgs/arch/PKGBUILD
Comment (Rafostar) [inline pkgs/arch/PKGBUILD:29]: python3 ?

File: pkgs/rpm/clapper.spec
Comment (Rafostar) [inline pkgs/rpm/clapper.spec:26]: One more tab before `noarch` for text align clarity.

**Review by Rafostar — CHANGES_REQUESTED** [review]

File: pkgs/rpm/clapper.spec
Comment (sp1ritCS) [reply pkgs/rpm/clapper.spec:29]: Ok, this seems specific to RHEL; thats why I didn't have any issues

File: pkgs/arch/.SRCINFO
Comment (sp1ritCS) [reply pkgs/arch/.SRCINFO:9]: meson depends on python, as you can see it builds on the buildservice

Comment (Rafostar) [reply pkgs/arch/.SRCINFO:9]: OK

File: pkgs/rpm/clapper.spec
Comment (Rafostar) [reply pkgs/rpm/clapper.spec:29]: @sp1ritCS 
Also I am not sure since I am not used to packaging like you. Should this be here or inside fedora/rhel requirements? Normally `gtkglsink` comes inside `gstreamer1-plugins-good` (according to GStreamer website), but in Fedora it is in separate package `gstreamer-plugins-good-gtk` (seems to be fedora/rhel specific).

Comment (sp1ritCS) [reply pkgs/rpm/clapper.spec:29]: `gtkglsink` is in `gstreamer*plugins-good-gtk` for all rpm based distros. Might be in debian or so the case that all of them are in `gstreamer*plugins-good`, but  only the rpm packages provide gstreamer1(element-gtkglsink)()(64bit).

https://pkgs.org/search/?q=gstreamer1(element-gtkglsink)()(64bit)

File: pkgs/rpm/clapper.spec
Comment (Rafostar) [inline pkgs/rpm/clapper.spec:29]: @sp1ritCS 
Like I noticed earlier `gstreamer` on Fedora/RHEL would install wrong (old) gstreamer-0.10. You should probably move it to `suse_version` and add `gstreamer1` inside the fedora/rhel. This is what Totem does.

File: pkgs/rpm/clapper.spec
Comment (Rafostar) [inline pkgs/rpm/clapper.spec:33]: What do you think about adding `gstreamer*-libav` to "Recommends" (and "optdepends" on Arch). They are not required, but the player won't be able to handle most media files (especially videos) without it so maybe we should recommend them at least.

File: pkgs/flatpak/com.github.rafostar.Clapper.yml
Comment (Rafostar) [inline pkgs/flatpak/com.github.rafostar.Clapper.yml:15]: > maybe add additinal codecs?

When you have time, please see my earlier comment about this [https://github.com/Rafostar/clapper/pull/3#issuecomment-691436334](https://github.com/Rafostar/clapper/pull/3#issuecomment-691436334). I would like to hear your opinion.

I **do not** ask you to work on them right now. I just would like to know your opinion before I merge and codecs can always be added later (the player does not work with flatpak right now, so this is not a high priority).

**Review by Rafostar — CHANGES_REQUESTED** [review]

File: pkgs/flatpak/com.github.rafostar.Clapper.yml
Comment (sp1ritCS) [reply pkgs/flatpak/com.github.rafostar.Clapper.yml:16]: That was the same Idea I had too, I just implemented it, wasn't that hard

File: pkgs/rpm/clapper.spec
Comment (sp1ritCS) [reply pkgs/rpm/clapper.spec:33]: If I see that right, then there is no official gstreamer1-libav for RHEL. Only RPM Fusion Free & UnitedRPMs.

I potential workaround would be to add gstreamer1-libav to my repo aswell

File: pkgs/flatpak/com.github.rafostar.Clapper.yml
Comment (Rafostar) [inline pkgs/flatpak/com.github.rafostar.Clapper.yml:19]: > gstreamer from the gnome runtime is outdated, maybe upgrade all modules?

Its outdated in 3.36 runtime. The 3.38 will have the latest version. We always have to make sure that modules we include match the gstreamer minor version from runtime. Otherwise it wont work. There is probably no need to compile more modules then necessary, as each time a new GNOME version is released the latest GStreamer appears in the runtime shortly afterwards.

File: pkgs/flatpak/com.github.rafostar.Clapper.yml
Comment (Rafostar) [reply pkgs/flatpak/com.github.rafostar.Clapper.yml:16]: Thanks

File: pkgs/flatpak/com.github.rafostar.Clapper.yml
Comment (sp1ritCS) [reply pkgs/flatpak/com.github.rafostar.Clapper.yml:19]: i know, I just didn't know if you depend on anything > 1.16.2. I used to correct gst-vaapi version

File: pkgs/rpm/clapper.spec
Comment (Rafostar) [reply pkgs/rpm/clapper.spec]: Ah. Yes. You are right. I forgot Fedora ships them in rpmfusion (proprietary code). I will simply update README to tell people who wish to install/test this player from source code to install them separately.

> would be to add gstreamer1-libav to my repo aswell

The decision is up to you.

File: pkgs/flatpak/com.github.rafostar.Clapper.yml
Comment (Rafostar) [reply pkgs/flatpak/com.github.rafostar.Clapper.yml:19]: The gstreamer 1.18 was just released few days ago. As of now all code was made using 1.16. I do not know if 3.38 + 1.18 runtime is already available. We can always update the `.yml` file later.

File: pkgs/rpm/clapper.spec
Comment (sp1ritCS) [reply pkgs/rpm/clapper.spec]: > I potential workaround would be to add gstreamer1-libav to my repo aswell

Not gonna happen, due to legal problems with MPEG-LA. [Software which is encumbered by patent claims](https://en.opensuse.org/openSUSE:Build_Service_application_blacklist#Software_which_is_encumbered_by_patent_claims).

I think the best way is to just tell RHEL users to install gstreamer1-libav from RPM Fusion Free if they want additional codecs.

**Review by Rafostar — APPROVED** [review]
Seems to be all good now. Are you still tweaking something or can I merge?

Also do you plan to create some repo besides the "testing" one, that I could add to README and tell people who wish to test this player to add it to their package manager?

**Comment by sp1ritCS** [comment]
You should be ready to merge now.

About the repository, I could submit it to my main repo, where It'll get built for
* openSUSE
  * Tumbleweed
  * Leap
    * 15.1
    * 15.2
* SuSE Enterprise Linux
  * 15
  * 15 SP2
* Fedora
  * 32
  * Rawhide

Or I can create a sub repository, that could contain multiple packages. That would have multiple benefits, first you could use it as a centralized place to build [**all**](https://user-images.githubusercontent.com/31540351/93006649-680a9680-f54e-11ea-9af8-4bf14dc9ac41.png) packages, not just suse / fedora.
As you currently do not have CI, you could use the OBS to automatically build the latest commit, on push to master in a sub-repo of this. Similar to `pkgname` & `pkgname-git` on the AUR.

It's up to you if you want/need CI.

**Comment by Rafostar** [comment]
> It's up to you if you want/need CI.

Since this is done in GJS, the CI does not give much benefit here. The project does not need any code compiling, so the builds by CI would always succeed (even after bad commit).

Once I consider this project good enough to tag a new release, I will try to publish (and maintain) this on flathub. I will leave all other eventual packages to maintainers (like you) interested in this project. You can do an unstable (git) version packaging if you want, but I think that maintaining stable releases (once we have one) is enough.

When (wait for stable or do unstable) and where you want to build it, is entirely up to you. If you decide to host a repo that will contain the builds for download (or adding to package manager), just send me a PR with updated README with the links.

**Comment by resynth1943** [comment]
You could use TypeScript for that.
Also, can't wait to play with the Flatpak ;D

**Pull request merged**
